examples/src/test/java/io/grpc/examples/header/HeaderServerInterceptorTest.java

```diff
@@ -0,0 +1,135 @@
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Server;
+import io.grpc.ServerInterceptors;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterBlockingStub;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterImplBase;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.StreamObserver;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+
+import java.io.IOException;
+
+/**
+ * Unit tests for {@link HeaderClientInterceptor}.
+ * For demonstrating how to write gRPC unit test only.
+ * Not intended to provide a high code coverage or to test every major usecase.
+ *
+ * <p>For basic unit test examples see {@link io.grpc.examples.helloworld.HelloWorldClientTest} and
+ * {@link io.grpc.examples.helloworld.HelloWorldServerTest}.
+ */
+@RunWith(JUnit4.class)
+public class HeaderServerInterceptorTest {
+  private Server fakeServer;
+  private ManagedChannel inProcessChannel;
+  private ClientCall.Listener<?> spyListener;
+
+  @Before
+  public void setUp() throws IOException {
+    String uniqueServerName = "fake server for " + getClass();
+    GreeterImplBase greeterImplBase =
+        new GreeterImplBase() {
+          @Override
+          public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
+            responseObserver.onNext(HelloReply.getDefaultInstance());
+            responseObserver.onCompleted();
+          }
+        };
+    fakeServer = InProcessServerBuilder.forName(uniqueServerName)
+        .addService(ServerInterceptors.intercept(greeterImplBase, new HeaderServerInterceptor()))
+        .build()
+        .start();
+    inProcessChannel = InProcessChannelBuilder.forName(uniqueServerName).build();
+  }
+
+  @After
+  public void tearDown() {
+    inProcessChannel.shutdownNow();
+    fakeServer.shutdownNow();
+  }
+
+  @Test
+  public void serverHeaderDeliveredToClient() {
+    ClientInterceptor clientInterceptor =
+        new ClientInterceptor() {
+          @Override
+          public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+              MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+            return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+              @SuppressWarnings("unchecked")
+              @Override
+              public void start(Listener<RespT> responseListener, Metadata headers) {
+                spyListener = spy(responseListener);
```

Comment:
This variable seems local to the test. It seems it was just convenient to use a class field. You may consider instead using an AtomicReference or named class:
```
public void serverHeaderDeliveredToClient() {
  class SpyingClientInterceptor implements ClientInterceptor {
    ClientCall.Listener<?> spyListener;
    ...
  }

  SpyingClientInterceptor clientInterceptor = new SpyingClientInterceptor();
  ...
  assertNotNull(clientInterceptor.spyListener);
```